### PR TITLE
Tie allowed network modes to namespace capabilities.

### PR DESCRIPTION
**Pull request by apollo13**
@tgross How does that look?

### CODE REVIEW

File: nomad/job_endpoint_validators.go
Comment (apollo13) [inline nomad/job_endpoint_validators.go:83]: Question to reviewers: Is this needed or should the NetworkResource always have a proper mode set? I know that the default is "host" but is that already set here?

Comment (tgross) [reply nomad/job_endpoint_validators.go:83]: We'll have passed thru the `(*NetworkResource).Canonicalize` method by this point, but that doesn't set the default mode. IIRC because of a backwards compatibility concern in the past... probably worth revisiting but what you've got here is fine.

File: nomad/job_endpoint_validators_test.go
Comment (tgross) [inline nomad/job_endpoint_validators_test.go:221]: We've been slowly migrating to [`shoenig/test`](https://pkg.go.dev/github.com/shoenig/test) which uses generics to fix some correctness issues in edge cases.

```suggestion
	must.EqError(t, err, "used group network mode \"host\" is not allowed in namespace \"default\"")
```

File: nomad/job_endpoint_validators_test.go
Comment (tgross) [inline nomad/job_endpoint_validators_test.go:225]:
```suggestion
	must.NoError(t, err)
```

File: nomad/job_endpoint_validators_test.go
Comment (tgross) [inline nomad/job_endpoint_validators_test.go:230]:
```suggestion
	must.EqError(t, err, "used group network modes [\"host\" \"cni/forbidden\"] are not allowed in namespace \"default\"")
```

File: nomad/job_endpoint_validators_test.go
Comment (tgross) [inline nomad/job_endpoint_validators_test.go:215]:
```suggestion
	must.NoError(t, s1.fsm.State().UpsertNamespaces(1000, []*structs.Namespace{ns}))
```

File: nomad/job_endpoint_validators.go
Comment (tgross) [inline nomad/job_endpoint_validators.go:58]: We end up having to repeat this logic in the function. Maybe change the signature of `taskValidateNetworkMode` so that it returns the mode name it checked?

File: nomad/job_endpoint_validators_test.go
Comment (tgross) [inline nomad/job_endpoint_validators_test.go:198]:
```suggestion
		must.Eq(t, c.result, taskValidateNetworkMode(network, c.ns), must.Sprint(c.description))
```

**Review by tgross — CHANGES_REQUESTED** [review]
This looks great @apollo13! I've left some minor comments. We'll also need:
* Can you run `make cl` to add a changelog entry
* Can you update the namespace specification docs as well? https://developer.hashicorp.com/nomad/docs/other-specifications/namespace

Just a heads up that I'm going to be out on vacation starting next week until September. I'll make sure someone else on the team has their eyes on this PR. But at worst I'll pick it back up and that'll be in time for Nomad 1.8.4.

**Comment by apollo13** [comment]
>     * Can you run `make cl` to add a changelog entry

Had one but fixed it up

>     * Can you update the namespace specification docs as well? https://developer.hashicorp.com/nomad/docs/other-specifications/namespace

done

> Just a heads up that I'm going to be out on vacation starting next week until September. I'll make sure someone else on the team has their eyes on this PR. But at worst I'll pick it back up and that'll be in time for Nomad 1.8.4.

Have fun :)

File: nomad/job_endpoint_validators_test.go
Comment (tgross) [inline nomad/job_endpoint_validators_test.go:143]: I added this test case just for thoroughness.

**Review by tgross — APPROVED** [review]
LGTM! Once CI is done I'll get this merged.

**Pull request merged**
